### PR TITLE
HomeWork

### PR DESCRIPTION
**Pull request by Yam0r**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,102 @@
...
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
+        this.size = 0;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:15]: do we need `this` here? [🤔](https://stackoverflow.com/questions/2411270/when-should-i-use-this-in-a-class/2411283#2411283)
size will be 0 anyway, 0 is default value for int data type

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,102 @@
...
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:20]: don't pass any args to this method, size is accessible in its scope

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,102 @@
...
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(size + list.size());
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:35]:
```suggestion
```
this logic is already in `add()`, this is redundant

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,102 @@
...
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null; // Clear the last element
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:58]:
```suggestion
        elements[--size] = null;
```
don't forget to remove your comments

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,102 @@
...
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && elements[i] == null) || (element != null
+                    && element.equals(elements[i]))) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:66]:
```suggestion
            if (element == elements[i] 
                   || element != null && element.equals(elements[i])) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,102 @@
...
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:70]: include element in the message, will make debugging easier

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,102 @@
...
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+    }
+
+    private void validateIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index for add operation: "
+                    + index);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:99]: include `size` in the message too

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private Object[] elements;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:10]: that's why we have class type parameter here, to not use non-specific `Object` here

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job overall!

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,100 @@
...
+    private int size;
+
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:15]:
```suggestion
        elements = new Object[DEFAULT_CAPACITY];
```
not fixed

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,100 @@
...
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private Object[] elements;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:11]:
```suggestion
    private T[] elements;
```
not fixed

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,100 @@
...
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, elements[i])) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:64]: I know that IDEA suggests to change to Objects.equals, but it's forbidden by task requirements
https://github.com/mate-academy/jv-array-list/blob/master/checklist.md#dont-use-objects-arrays-or-any-other-util-class

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,48 +1,100 @@
...
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > elements.length) {
+            int newCapacity = Math.max((int) (elements.length * GROWTH_FACTOR), minCapacity);
+            elements = Arrays.copyOf(elements, newCapacity);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:84]: https://github.com/mate-academy/jv-array-list/blob/master/checklist.md#dont-use-objects-arrays-or-any-other-util-class

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request closed without merging**